app/Main.hs

```diff
@@ -0,0 +1,352 @@
+    -- ^ Number of digits in a password
+    TotpMk Algorithm RawSharedSecret Time Timespan Digits
+  | -- | Check presented password against time periods.
+    -- ^ Shared secret
+    -- ^ Valid counter range, before and after ideal
+    -- ^ Time of TOTP
+    -- ^ Time range in seconds
+    -- ^ Numer of digits in a password
+    -- ^ Password given by user
+    TotpCheck Algorithm RawSharedSecret (Word64, Word64) Time Timespan Digits Text
+  | -- | Create a URI suitable for authenticators.
+    -- ^ Shared secret key. Must be encoded in base32.
+    -- ^ Name of the account (usually an email address)
+    -- ^ Issuer
+    -- ^ Amount of digits expected from the end-user
+    -- ^ Amount of time before the generated code expires
+    -- ^ Algorithm required
+    TotpMkUri Text Text Text Digits Timespan Algorithm
+  | -- | Create an new random key to be used with the SHA-1 functions
+    KeyGen Algorithm
+
+cmdP :: Time -> Parser Cmd
+cmdP now' =
+  hsubparser
+    ( command "hotp" (info hotpCmdP (progDesc "HOTP commands"))
+        <> command "totp" (info totpCmdP (progDesc "TOTP commands"))
+        <> command "key-gen" (info keyGenCmdP (progDesc "Generate keys"))
+    )
+  where
+    hotpCmdP =
+      hsubparser $
+        command
+          "make"
+          (info hotpMakeP (progDesc "Compute HMAC-Based One-Time Password using secret key and counter value"))
+          <> command
+            "check"
+            (info hotpCheckP (progDesc "Check presented password against a valid range"))
+
+    totpCmdP =
+      hsubparser $
+        command
+          "make"
+          (info totpMakeP (progDesc "Compute a Time-Based One-Time Password using secret key and time"))
+          <> command
+            "check"
+            (info totpCheckP (progDesc "Check presented password against time periods"))
+          <> command
+            "uri"
+            (info totpUriP (progDesc "Create a URI suitable for authenticators"))
+
+    hotpMakeP :: Parser Cmd
+    hotpMakeP =
+      HotpMk
+        <$> algorithmP
+        <*> rawSharedSecretP
+        <*> counterP
+        <*> digitsP
+
+    hotpCheckP :: Parser Cmd
+    hotpCheckP =
+      HotpCheck
+        <$> algorithmP
+        <*> rawSharedSecretP
+        <*> timeRangeP
+        <*> counterP
+        <*> digitsP
+        <*> passwordP
+
+    totpMakeP :: Parser Cmd
+    totpMakeP =
+      TotpMk
+        <$> algorithmP
+        <*> rawSharedSecretP
+        <*> timeP
+        <*> timespanP
+        <*> digitsP
+    totpCheckP :: Parser Cmd
+    totpCheckP =
+      TotpCheck
+        <$> algorithmP
+        <*> rawSharedSecretP
+        <*> timeRangeP
+        <*> timeP
+        <*> timespanP
+        <*> digitsP
+        <*> passwordP
+    totpUriP :: Parser Cmd
+    totpUriP =
+      TotpMkUri
+        <$> base32RawSharedSecretP
+        <*> accountNameP
+        <*> issuerP
+        <*> digitsP
+        <*> timespanP
+        <*> algorithmP
+
+    keyGenCmdP :: Parser Cmd
```

Comment:
Unless `Command` is already used by optparse-applicative, I'd rather you used it instead of `Cmd`.